pkg/kubelet/prober/runner.go

```diff
@@ -0,0 +1,335 @@
+	err := hp.sync(container, status, probeType)
+	if err != nil {
+		return probe.Unknown, "", err
+	}
+
+	klog.V(4).InfoS("HTTP-Probe", "scheme", hp.httpRequest.URL.Scheme, "host", hp.host,
+		"port", hp.httpRequest.URL.Port(), "path", hp.httpRequest.URL.Path, "timeout", hp.timeout,
+		"headers", hp.httpRequest.Header)
+	return prober.http.Probe(hp.httpRequest, hp.timeout)
+}
+
+// grpcProbeRunner holds state and logic for running GRPC probes.
+type grpcProbeRunner struct {
+	host    string
+	service string
+	port    int
+	timeout time.Duration
+}
+
+// newGRPCProbeRunner returns grpcProbeRunner which implements probeRunner.
+func newGRPCProbeRunner() *grpcProbeRunner {
+	return &grpcProbeRunner{}
+}
+
+func (gp *grpcProbeRunner) sync(container v1.Container, status v1.PodStatus, probeType probeType) error {
+	if gp.host == status.PodIP {
+		// no sync required, podIP hasn't changed.
+		return nil
+	}
+
+	// no need to handle error here, factory will handle this case.
+	probeSpec, _ := getProbeSpecFromContainer(container, probeType)
+
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+
+	host := status.PodIP
+	service := ""
+	if probeSpec.GRPC.Service != nil {
+		service = *probeSpec.GRPC.Service
+	}
+
+	// persist
+	gp.host = host
+	gp.service = service
+	gp.port = int(probeSpec.GRPC.Port)
+	gp.timeout = timeout
+	return nil
+}
+
+func (gp *grpcProbeRunner) run(_ context.Context, container v1.Container, status v1.PodStatus, probeType probeType, prober *prober) (probe.Result, string, error) {
+	err := gp.sync(container, status, probeType)
+	if err != nil {
+		return probe.Unknown, "", err
+	}
+
+	klog.V(4).InfoS("GRPC-Probe", "host", gp.host, "service", gp.service, "port", gp.port, "timeout", gp.timeout)
+	return prober.grpc.Probe(gp.host, gp.service, gp.port, gp.timeout)
+}
+
+// execProbeRunner holds state and logic for running Exec probes.
+type execProbeRunner struct {
+	synced      bool
+	containerID kubecontainer.ContainerID
+	command     []string
+	timeout     time.Duration
+}
+
+// newExecProbeRunner returns execProbeRunner which implements probeRunner.
+func newExecProbeRunner() *execProbeRunner {
+	return &execProbeRunner{}
+}
+
+func (ep *execProbeRunner) sync(container v1.Container, status v1.PodStatus, probeType probeType) error {
+	// we only need to sync execProberRunner once as it doesn't depend on podIP and all other
+	// required objects are immutable, not using sync.Once here to avoid un-necessary mutexes.
+	if ep.synced {
+		return nil
+	}
+
+	// no need to handle error here, factory will handle this case.
+	probeSpec, _ := getProbeSpecFromContainer(container, probeType)
+
+	command := kubecontainer.ExpandContainerCommandOnlyStatic(probeSpec.Exec.Command, container.Env)
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+
+	c, _ := podutil.GetContainerStatus(status.ContainerStatuses, container.Name)
+	containerID := kubecontainer.ParseContainerID(c.ContainerID)
```

Comment:
I wonder why we didn't do this in the past, instead of passing around `containerID`... are we sure we won't suffer from status propagation delays on container init?

Reply:
Also it's only ever used for logs, and just for one type of probe??? Weird, but maybe worth doing some archaeological work to make sure...